python-client/hatchet/workflow.py

```diff
@@ -0,0 +1,110 @@
+from .client import new_client 
+from .workflows_pb2 import CreateWorkflowVersionOpts, CreateWorkflowJobOpts, CreateWorkflowStepOpts
+from dotenv import load_dotenv
+from typing import Callable, List, Tuple, Any
+
+# TODO: don't load dotenv
+load_dotenv()
+
+def workflow(name : str='', on_events : list=[], on_crons : list=[]):
+   def inner(cls):
+        cls.on_events = on_events
+        cls.on_crons = on_crons
+        cls.name = name or str(cls.__name__)
+
+        # Define a new class with the same name and bases as the original, but with WorkflowMeta as its metaclass
+        return WorkflowMeta(cls.__name__, cls.__bases__, dict(cls.__dict__))
+   
+   return inner
+
+def step(name : str='', parents : List[str] = []):
+    def inner(func):
+        func._step_name = name or func.__name__
+        func._step_parents = parents
+
+        # print length of parents
+        print("len(parents)", len(parents), len(func._step_parents))
+
+        return func
+
+    return inner
+
+# def step(func):
+#     if not hasattr(func, 'name'):
+#         func._step_name = func.__name__
+#     else:
+#         func._step_name = func.name
+
+#     if not hasattr(func, 'parents'):
+#         func._step_parents = []
+#     else:
+#         func._step_parents = func.parents
+
+#     return func
+    
+
+# Initialize the counter for ordering steps
+step.order_counter = 0
+
+stepsType = List[Tuple[str, Callable[..., Any]]]
+
+class WorkflowMeta(type):
```

Comment:
what is the point of this?

Reply:
I wasn't sure of a better/more pythonic way of tracking the attributes of the workflow class, like the steps, without using a metaclass, along with creating the workflow over the API when the child class is instantiated, which is why I modified `__new__`. As far as I understand, one use-case for metaclasses is adding classes/attributes of the class to separate data structures and potentially modifying the behavior of the child class. This could probably be done with a decorator function that modifies the class directly, I'm not sure which approach is better though.